pageserver/src/basebackup.rs

```diff
@@ -252,6 +254,20 @@ where
     async fn send_tarball(mut self) -> Result<(), BasebackupError> {
         // TODO include checksum
 
+        // Detect if we are creating the basebackup exactly at a shutdown checkpoint.
+        let normal_shutdown =
+            if let Ok(checkpoint_bytes) = self.timeline.get_checkpoint(self.lsn, self.ctx).await {
+                let checkpoint =
+                    CheckPoint::decode(&checkpoint_bytes).context("deserialize checkpoint")?;
+                let checkpoint_end_lsn = calculate_walrecord_end_lsn(
+                    Lsn(checkpoint.redo),
+                    XLOG_SIZE_OF_XLOG_RECORD + 8 + SIZEOF_CHECKPOINT,
```

Comment:
Where does the `+ 8` come from?

Reply:
Sorry, doesn't know how to name it. It aligned encoded length of main data:
```
	/* followed by main data, if any */
	if (mainrdata_len > 0)
	{
		if (mainrdata_len > 255)
		{
			uint32		mainrdata_len_4b;

			if (mainrdata_len > PG_UINT32_MAX)
				ereport(ERROR,
						(errmsg_internal("too much WAL data"),
						 errdetail_internal("Main data length is %llu bytes for a maximum of %u bytes.",
											(unsigned long long) mainrdata_len,
											PG_UINT32_MAX)));

			mainrdata_len_4b = (uint32) mainrdata_len;
			*(scratch++) = (char) XLR_BLOCK_ID_DATA_LONG;
			memcpy(scratch, &mainrdata_len_4b, sizeof(uint32));
			scratch += sizeof(uint32);
		}
		else
		{
			*(scratch++) = (char) XLR_BLOCK_ID_DATA_SHORT;
			*(scratch++) = (uint8) mainrdata_len;
		}
		rdt_datas_last->next = mainrdata_head;
		rdt_datas_last = mainrdata_last;
		total_len += mainrdata_len;
	}
```

Reply:
Hmm, so it's `XLogRecordDataHeaderShort` ? But that's only 2 bytes

Reply:
We don't check explicitly that it was a shutdown checkpoint. (Because we don't store that information.) Is it possible that this would get fooled by an online checkpoint, if there was no activity during the checkpoint? Ie. is it possible that the redo pointer of an online checkpoint would also point to the beginning of the checkpoint WAL record?

Starting with v17, an online checkpoint always creates a XLOG_CHECKPOINT_REDO record, and the redo pointer always points to that. But I don't think there's any hard guarantee on previous versions. The XLOG_RUNNING_XACTs record is created too, but that depends on wal_level.

Reply:
All in all, this feels very fiddly. Maybe it's good enough for now, the consequence if this goes wrong isn't _so_ bad, after all. But I wonder if it would be more straightforward to add the "is this a shutdown checkpoint" information to the persistent data during WAL ingestion, after all.

Please add comments explaining all these assumptions we make here.

Reply:
Actually it was your suggestion (although I also think about something like this from the very beginning):
```
I'm suggesting that we store the flag in the checkpoint key-value pair itself.
Concretely, something like this (this does not work as is, though, see below):

diff --git a/pageserver/src/basebackup.rs b/pageserver/src/basebackup.rs
index a32d09f3b..a8a30ca8c 100644
--- a/pageserver/src/basebackup.rs
+++ b/pageserver/src/basebackup.rs
@@ -678,6 +678,12 @@ async fn add_pgcontrol_file(&mut self) -> Result<(), BasebackupError> {
             self.timeline.pg_version,
         )?;
 
+        // If there was a shutdown checkpoint at this LSN, set the dbstate in the control
+        // file accordingly
+        if checkpoint.dbstate == DB_SHUTDOWNED {
+            pg_control.dbstate = DB_SHUTDOWNED;
+        }
+
         //send pg_control
         let header = new_tar_header("global/pg_control", pg_control_bytes.len() as u64)?;
         self.ar
diff --git a/pageserver/src/walingest.rs b/pageserver/src/walingest.rs
index 229c01a68..13d201b45 100644
--- a/pageserver/src/walingest.rs
+++ b/pageserver/src/walingest.rs
@@ -150,6 +150,14 @@ pub async fn ingest_record(
             self.checkpoint_modified = true;
         }
 
+        // Since we ingested a record, we are certainly not at the shutdowned state anymore.
+        // (If this record itself is another shutdown checkpoint record, we'll overwrite this
+        // back to DB_SHUTDOWNED again below.)
+        if self.checkpoint.dbstate == DB_SHUTDOWNED {
+            self.checkpoint.state == DB_IN_PRODUCTION;
+            self.checkpoint_modified = true;
+        }
+
         failpoint_support::sleep_millis_async!("wal-ingest-record-sleep");
 
         match decoded.xl_rmid {
@@ -475,6 +483,11 @@ pub async fn ingest_record(
                             cp.oldestActiveXid = xlog_checkpoint.oldestActiveXid;
                         }
 
+                        // Remember if we were in a shutdowned state at this LSN
+                        if info == pg_constants::XLOG_CHECKPOINT_SHUTDOWN {
+                            cp.state == DB_SHUTDOWNED;
+                        }
+
                         // Write a new checkpoint key-value pair on every checkpoint record, even
                         // if nothing really changed. Not strictly required, but it seems nice to
                         // have some trace of the checkpoint records in the layer files at the same

So this way, the state "was this a shutdown record LSN" is persisted in the chekpoint key-value pair. When you ingest a shutdown checkpoint record, you set the flag in the checkpoint struct, and persist it. When you ingest the first record after a shutdown checkpoint, you clear the flag in the checkpoint struct and pesrsist it. This is similar to how we update the nextXid for example.

Now, that doesn't quite work: there is no 'state' field in the Checkpoint struct. We'd need to store that somewhere else. We do persists the control file in a key-value pair too, so we could update that. We currently never update the control file after the initial import, but there's no fundamental reason for that, we just haven't needed to do that.
So, instead of updating the 'state' in the CheckPoint struct in that patch, update it in the ControlFileData struct. I think that means keeping the latest version of ControlFileData in memory in WalIngest, just like it keeps the latest CheckPoint.

After writing all that, it occurs to me: could we look at the existing CheckPoint.redo field in [basebackup.rs](http://basebackup.rs/) instead, and get away with no changes to [walingest.rs](http://walingest.rs/) ? If it 'redo' points to the LSN that you're taking the basebackup from, then you're starting from shutdown checkpoint. Otherwise it's not. Not sure if that quite works; I think the redo pointer points to the beginning of the shutdown record while the basebackup LSN is just after the shutdown record. But maybe there's something you could do there?


Maybe check if (distance between CheckPoint.redo and the basebackup LSN) == sizeof(CheckPoint). Taking into account if it's split across pages correctly.
```

I wonder if you put all this into comment:)

Concerning persisting information that we have observed SHUTDOWN checkpoint and clearing it after ingesting next WAL statement: you know my opinion. I think that it will be better to use deteriorated statistic rather than drop statistic at all and have problems with vacuum. Still not clear how critical this problem can be. But it is known=real problem. In case of lack of statistic, table is skipped in vacuum. The problems with using deteriorated statistic are obscure. Yes, it can contain entries for dead (dropped)_ relations. But statistic is not updated immediately in any case: if some relation is dropped it will not cause immediate update of statistic. 

So it seems to me to be strange to spend much efforts just to preserve current  behaviour of Postgres which I consider more than bug rather than feature and which most likely will be changed soon...

Looks like the only real alternative (if we do not want to change database format) is to persist information about shutdown checkpoint in control file. And clear it at any subsequent record. So we need to add one more  mutex `Mutex<ControlFile>` to Timeline, initialize it on timeline start, update on each checkpoint and check on each ingested WAL record. It will cause creation of 2*N versions of control files when N is number of shutdowns.
I do not expect that it will somehow affect performance and storage size. But still it is IMHO useless stupid job. 

Current solution is much simpler and so less error prone. May be there is very small probability that in some older version of Postgres (15 or 16) we wrongly interpret online checkpoint as shutdown checkpoint and include pgstat information is basebackup while it is actually deteriorated. Do you really some risk here?

Reply:
> Hmm, so it's `XLogRecordDataHeaderShort` ? But that's only 2 bytes

Yes. May be it is better to replace 8 with 2 or `XLogRecordDataHeaderShort`
Real size of shutdown checkpoint record is 114 bytes and doe to 8 bytes alignment it is aligned to 120.